### ISSUE TITLE
Implement ISO_Fortran_binding.h

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#1966]
The Fortran 2018 standard implements ISO_Fortran_binding.h. Now when we have physical arrays, it's the `ISODescriptorArray` of `array_physical_type`, so we just have to provide casts, and then some C header file I think.

**Comment by certik** [comment]
Yes, we handle that by the new [ArrayPhysicalCast](https://github.com/lfortran/lfortran/blob/4a8f2eb04408f6623ad58d02e39b65a4c51ef1b2/src/libasr/ASR.asdl#L335) node. We already do this for pointer arrays vs descriptor arrays and now we just add support for `ISODescriptorArray`.

**Comment by awvwgk** [comment]
For this it could be useful to consolidate the headers in GFortran, LLVM flang, Intel, NAG and co. and decide whether there is an ABI for the CFI descriptors LFortran should be compatible with.

**Comment by ivan-pi** [comment]
Great point. Given LFortran can reuse gfortran module files, it would make sense to have compatibility here.

**Comment by certik** [comment]
Is Flang binary compatible with GFortran for this ISO descriptor?

**Comment by ivan-pi** [comment]
Short answer is NO, based on the layout of the descriptor struct:

The flang descriptor is:
```c
/* 18.5.3 generic data descriptor */
typedef struct CFI_cdesc_t {
  /* These three members must appear first, in exactly this order. */
  void *base_addr;
  size_t elem_len; /* element size in bytes */
  int version; /* == CFI_VERSION */
  CFI_rank_t rank; /* [0 .. CFI_MAX_RANK] */
  CFI_type_t type;
  CFI_attribute_t attribute;
  unsigned char f18Addendum;
#ifdef __cplusplus
  cfi_internal::FlexibleArray<CFI_dim_t> dim;
#else
  CFI_dim_t dim[]; /* must appear last */
#endif
} CFI_cdesc_t;
```
whereas the `gfortran` descriptor is
```c
typedef struct CFI_cdesc_t
 {
    void *base_addr;
    size_t elem_len;
    int version;
    CFI_rank_t rank;
    CFI_attribute_t attribute;
    CFI_type_t type;
    CFI_dim_t dim[];
 }
CFI_cdesc_t;
```

The NAG descriptor is:
```c
typedef struct {
  void *base_addr;
  size_t elem_len;
  int version;
  CFI_rank_t rank;
  CFI_type_t type;
  CFI_attribute_t attribute;
  size_t len,offset;
  int64_t *dtp;
  void *sig;
  void *reserved;
  int64_t ireserved;
  CFI_dim_t dim[];
} CFI_cdesc_t;
```

and the Intel Fortran descriptor is:
```c
typedef struct CFI_cdesc_t_
{
    void * base_addr;
    size_t elem_len;
    int version;
    CFI_attribute_t attribute;
    CFI_rank_t rank;
    CFI_type_t type;
    // Begin Intel-specific fields - these are for
    // Intel compiler/library use only and should 
    // not be modified
    intptr_t intel_flags;
    intptr_t intel_reserved1;
    intptr_t intel_reserved2;
    // End Intel-specific fields
    CFI_dim_t dim[];
} CFI_cdesc_t;
```

**Comment by ivan-pi** [comment]
The differences are more visible if we align these:

```c
// gfortran                 // flang                    // NAG                      // Intel         
void *base_addr;            void *base_addr;            void *base_addr;            void * base_addr;
size_t elem_len;            size_t elem_len;            size_t elem_len;            size_t elem_len;
int version;                int version;                int version;                int version;
CFI_rank_t rank;            CFI_rank_t rank;            CFI_rank_t rank;            CFI_attribute_t attribute;
CFI_attribute_t attribute;  CFI_type_t type;            CFI_type_t type;            CFI_rank_t rank;
CFI_type_t type;            CFI_attribute_t attribute;  CFI_attribute_t attribute;  CFI_type_t type;
CFI_dim_t dim[];            unsigned char f18Addendum;  size_t len,offset;          intptr_t intel_flags;
                            CFI_dim_t dim[];            int64_t *dtp;               intptr_t intel_reserved1;
                                                        void *sig;                  intptr_t intel_reserved2;
                                                        void *reserved;             CFI_dim_t dim[];
                                                        int64_t ireserved;
                                                        CFI_dim_t dim[];
```

Probably the typedefs `CFI_rank_t`, `CFI_attribute_t`, ..., are also of different width among implementations. 

The ABI compatibility is discussed in https://github.com/j3-fortran/fortran_proposals/issues/247 and the [Discourse thread](https://fortran-lang.discourse.group/t/is-iso-fortran-binding-h-compatible-for-different-fortran-abis/2652/7).

**Comment by certik** [comment]
Thanks a lot for this @ivan-pi ! Given all this, which descriptor would you recommend we use? It looks like they use all kinds of permutations of (rank, type, attribute). After that are compiler specific attributes, and then finally the "dim".

**Comment by ivan-pi** [comment]
The position of the first three struct members (`base_addr`, `elem_len`, and `version`) is prescribed by the standard. The `dim` member must come last.

That's a multi-faceted question. The vendor-specific attributes are probably meaningless because we don't know what purpose they serve. You can replace them with padding.

In the threads mentioned, Peter Klausler, Jeff Hammond, and Steve Lionel, had the opinion that it's too late to achieve ABI compatibility, because vendors are unlikely to change, and Fortran modules are incompatible anyway, so we usually compile from source. In HPC it's still common to build from source (e.g. with Spack) for performance tuning reasons, so ABI problems matter less I guess. Our center for examples provides modules compiled with Intel and GCC. (We have Intel Skylake Xeon.)
Also in the engineering and scientific industries that use Fortran, I think it's common to stick to one compiler for shipping the final software app. 

On personal systems, users like to install software in binary format and forget about it. A stable ABI is then an advantage, because you can rely on your system package manager. So on most Linux distros this would be GCC; that also applies to Python packages using Fortran. I guess most scientific Python users don't install the Intel OneAPI toolkit or the NVIDIA HPC SDK. NAG is out of the equation for non-institutional clients.

So my suggestion would be to seek for symbiosis with gfortran or flang, if you think that makes sense.

**Comment by certik** [comment]
Presumably NVIDIA and AMD will build their compilers on top of Flang, so choosing the Flang descriptor might make the most sense. Yes, mod file compatibility is tough, but `bind(c)` is binary compatible, as far as I know. And this is part of `bind(c)`, so it's a missed opportunity that they didn't make the descriptor binary compatible, but maybe it's not possible.

**Comment by ivan-pi** [comment]
I'm not sure where AMD stands currently. The AMD ROCm repository contains a fork of _new_ flang including the Fortran binding header [here](https://github.com/RadeonOpenCompute/llvm-project/blob/rocm-5.6.0/flang/include/flang/ISO_Fortran_binding.h). However the documents for [AOCC](https://www.amd.com/en/developer/aocc.html) (AMD Optimizing Compiler for C/C++, and Fortran) state that `amdflang` is based on [classic flang](https://github.com/flang-compiler/flang). Some documents from ORNL mention that `amdflang` is not really production ready (but on Frontier they have the Cray Compiler Environment anyways).

I checked the NVIDIA HPC SDK 23.3 and it already contains the flang header. The opportunity which would open up here would be to use `nvcc`/`nvc++` for GPU kernels.
